// ngx-show-hide-password/esm5/lib/show-hide-input.directive.js
import { __decorate } from "tslib";
import { Directive, ElementRef, Renderer2, OnInit, OnDestroy } from '@angular/core';
import { ShowHideService } from './show-hide.service';
var ShowHideInputDirective = /** @class */ (function () {
    function ShowHideInputDirective(service, el, renderer) {
        this.service = service;
        this.el = el;
        this.renderer = renderer;
        this.id = this.el.nativeElement.getAttribute('id');
        if (!this.id) {
            throw new Error("No attribute [id] found.");
        }
        this.service.setShow(this.id, this.el.nativeElement.type !== 'password');
    }
    ShowHideInputDirective.prototype.ngOnInit = function () {
        var _this = this;
        this.service
            .getObservable(this.id)
            .subscribe(function (show) {
            return _this.renderer.setAttribute(_this.el.nativeElement, 'type', show ? 'text' : 'password');
        });
    };
    ShowHideInputDirective.prototype.ngOnDestroy = function () {
        if (this.subscription) {
            this.subscription.unsubscribe();
        }
    };
    ShowHideInputDirective.ctorParameters = function () { return [
        { type: ShowHideService },
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    ShowHideInputDirective = __decorate([
        Directive({
            selector: 'input[showHideInput]'
        })
    ], ShowHideInputDirective);
    return ShowHideInputDirective;
}());
export { ShowHideInputDirective };
