// ngx-show-hide-password/esm5/lib/show-hide.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { ReplaySubject } from 'rxjs';
import * as i0 from "@angular/core";
var ShowHideService = /** @class */ (function () {
    function ShowHideService() {
        this.states = [];
    }
    ShowHideService.prototype.getIO = function (id) {
        var io = this.states.find(function (o) { return o.id === id; });
        if (!io && id) {
            io = this.init(id);
        }
        return io;
    };
    ShowHideService.prototype.init = function (id) {
        var subject = new ReplaySubject(1);
        var io = {
            id: id,
            show: false,
            subject: subject
        };
        this.states.push(io);
        return io;
    };
    ShowHideService.prototype.saveAndProadcast = function (io, show) {
        io.show = show;
        io.subject.next(io.show);
    };
    ShowHideService.prototype.getObservable = function (id) {
        return this.getIO(id).subject;
    };
    ShowHideService.prototype.setShow = function (id, show) {
        this.saveAndProadcast(this.getIO(id), show);
    };
    ShowHideService.prototype.toggleShow = function (id) {
        var io = this.getIO(id);
        this.saveAndProadcast(io, !io.show);
    };
    ShowHideService.ɵprov = i0.ɵɵdefineInjectable({ factory: function ShowHideService_Factory() { return new ShowHideService(); }, token: ShowHideService, providedIn: "root" });
    ShowHideService = __decorate([
        Injectable({
            providedIn: 'root'
        })
    ], ShowHideService);
    return ShowHideService;
}());
export { ShowHideService };
