// ngx-show-hide-password/esm5/lib/show-hide-status.directive.js
import { __assign, __decorate } from "tslib";
import { Directive, ElementRef, Renderer2, Input, ErrorHandler, OnDestroy } from '@angular/core';
import { ShowHideService } from './show-hide.service';
var ShowHideStatusDirective = /** @class */ (function () {
    function ShowHideStatusDirective(service, el, renderer, errorHandler) {
        this.service = service;
        this.el = el;
        this.renderer = renderer;
        this.errorHandler = errorHandler;
    }
    Object.defineProperty(ShowHideStatusDirective.prototype, "showHideStatus", {
        set: function (config) {
            this.init(config);
        },
        enumerable: true,
        configurable: true
    });
    ShowHideStatusDirective.prototype.init = function (config) {
        var _this = this;
        var defaultConfig = {
            show: 'visibility',
            hide: 'visibility_off',
            materialIcon: false,
            id: null
        };
        this.config = __assign(__assign({}, defaultConfig), config);
        if (this.config.id) {
            this.subscription = this.service
                .getObservable(this.config.id)
                .subscribe(function (show) { return _this.updateStatus(show); });
        }
        else {
            this.errorHandler.handleError(new Error("Status can not be set without [id]."));
        }
    };
    ShowHideStatusDirective.prototype.updateStatus = function (show) {
        if (this.config.materialIcon) {
            this.renderer.setProperty(this.el.nativeElement, 'innerHTML', show ? this.config.hide : this.config.show);
        }
        else {
            this.renderer.removeClass(this.el.nativeElement, !show ? this.config.hide : this.config.show);
            this.renderer.addClass(this.el.nativeElement, show ? this.config.hide : this.config.show);
        }
    };
    ShowHideStatusDirective.prototype.ngOnDestroy = function () {
        if (this.subscription) {
            this.subscription.unsubscribe();
        }
    };
    ShowHideStatusDirective.ctorParameters = function () { return [
        { type: ShowHideService },
        { type: ElementRef },
        { type: Renderer2 },
        { type: ErrorHandler }
    ]; };
    __decorate([
        Input()
    ], ShowHideStatusDirective.prototype, "showHideStatus", null);
    ShowHideStatusDirective = __decorate([
        Directive({
            selector: '[showHideStatus]'
        })
    ], ShowHideStatusDirective);
    return ShowHideStatusDirective;
}());
export { ShowHideStatusDirective };
