// ngx-show-hide-password/esm2015/lib/show-hide-input.directive.js
import { __decorate } from "tslib";
import { Directive, ElementRef, Renderer2, OnInit, OnDestroy } from '@angular/core';
import { ShowHideService } from './show-hide.service';
let ShowHideInputDirective = class ShowHideInputDirective {
    constructor(service, el, renderer) {
        this.service = service;
        this.el = el;
        this.renderer = renderer;
        this.id = this.el.nativeElement.getAttribute('id');
        if (!this.id) {
            throw new Error(`No attribute [id] found.`);
        }
        this.service.setShow(this.id, this.el.nativeElement.type !== 'password');
    }
    ngOnInit() {
        this.service
            .getObservable(this.id)
            .subscribe(show => this.renderer.setAttribute(this.el.nativeElement, 'type', show ? 'text' : 'password'));
    }
    ngOnDestroy() {
        if (this.subscription) {
            this.subscription.unsubscribe();
        }
    }
};
ShowHideInputDirective.ctorParameters = () => [
    { type: ShowHideService },
    { type: ElementRef },
    { type: Renderer2 }
];
ShowHideInputDirective = __decorate([
    Directive({
        selector: 'input[showHideInput]'
    })
], ShowHideInputDirective);
export { ShowHideInputDirective };
